File: 백준/Gold/2632. 피자판매/피자판매.py
from collections import defaultdict
from pprint import pprint
def init(arr):
    length = len(arr) - 1

    ret = defaultdict(int)
    for y in range(1, length+1):
        sum = 0
        for x in range(length):
            nxt = length if (y+x) % length == 0 else (y+x) % length
            sum = sum + arr[nxt]
            ret[sum]+=1

    return ret

if __name__ == '__main__':
    target = int(input())
    N, M = map(int,input().split())
    A = [0]
    for i in range(N):
        A.append(int(input()))
    B = [0]
    for i in range(M):
        B.append(int(input()))

    mapA = init(A)
    mapB = init(B)
    # pprint(mapA)
    # pprint(mapB)
    maxA = max(mapA.keys())
    maxB = max(mapB.keys())
    mapA[maxA] = 1
    mapB[maxB] = 1

    ans = 0
    for key, value in mapA.items():
        if key == target:
           ans += value

        if key < target and ((target - key) in mapB):
            ans += mapA[key] * mapB[target - key]

    for key, value in mapB.items():
        if key == target:
            ans += value

    print(ans)

File: 백준/Gold/2632. 피자판매/피자판매.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	target = Integer.parseInt(br.readLine());
    	st = new StringTokenizer(br.readLine());
    	N = Integer.parseInt(st.nextToken());
    	M = Integer.parseInt(st.nextToken());
    	A = new int[N+1];
    	B = new int[M+1];
    	
    	for(int i = 1; i<=N; i++) {
    		A[i] = Integer.parseInt(br.readLine());
    	}
    	
    	for(int i = 1; i<=M; i++) {
    		B[i] = Integer.parseInt(br.readLine());
    	}
    	
    	
    	TreeMap<Integer,Integer> mapA = init(A);
    	TreeMap<Integer,Integer> mapB = init(B);
    	
    	
    	int maxA = mapA.lastKey();
    	int maxB = mapB.lastKey();
    	
    	mapA.put(maxA, 1);
    	mapB.put(maxB, 1);
    	
    	int ans = 0;
    	
    			
    			
    	for(int key : mapA.keySet()) {
    		if(key < target && mapB.containsKey(target - key)) {
    			ans += mapA.get(key) * mapB.get(target - key);
    		}
    	}
    	
    	
    	if (mapA.containsKey(target))
    		ans += mapA.get(target);
    	
    	if(mapB.containsKey(target))
    		ans += mapB.get(target);
    		
    	
    	System.out.println(ans);
    }
    
    
    static TreeMap<Integer, Integer> init(int arr[]) {
    	TreeMap<Integer, Integer> ret = new TreeMap<>();
    	int len = arr.length - 1;
    	
    	for(int y=1; y<=len; y++) {
    		int sum = 0;
    		for(int x= 0; x < len; x++) {
    			int nxt = (y + x) % len == 0 ? len : (y + x) % len;
    			sum += arr[nxt];
    			ret.put(sum, ret.getOrDefault(sum, 0) + 1);
    		}
    	}
    	
    	return ret;
    	
    }
//    
    static int target, N, M;
    static StringTokenizer st;
    static int A[], B[];
}